### ISSUE TITLE
Font Size Rounding and Decimals

### ISSUE THREAD

**Issue by teddympq** [koreader/koreader#11531]
I'm trying to set a precise font size, but 18 pt is slightly too small and 18.5 pt is slightly too big
I've noticed that setting the font size to 18.1 pt results the same as 18.5 pt. I'm not sure if this behavior is related to font rendering or if it's intended. It would be nice to be able to set decimal values for fonts. As a workaround I tried changing the UPM values for the fonts.

**Comment by poire-z** [comment]
The "size" you set via the UI is a bit of some abstract number - I'm not certain it relates to the unit "pt".

It goes through some adjustment (depending on your screen size and DPI) to reach the final size in real pixels given to crengine (our EPUB renderer):
Initial setting on book loading:
https://github.com/koreader/koreader/blob/6f896e9383008d1e9f07369f39b814aece3ff211/frontend/apps/reader/modules/readerfont.lua#L177
Later font size change:
https://github.com/koreader/koreader/blob/6f896e9383008d1e9f07369f39b814aece3ff211/frontend/apps/reader/modules/readerfont.lua#L235-L242

https://github.com/koreader/koreader-base/blob/5c7d8bc2b5ca15938880711f75ddc045251e9625/ffi/framebuffer.lua#L412-L424

In the end, it's a size in (integer, no fractional) pixels.
If you long-press on "Font Size", the popup tells you approximately what's the resulting font size in "pt" (whatever that means in our context, real screen size or related to what you set as DPI, no idea).
![image](https://github.com/koreader/koreader/assets/24273478/ef969c4d-b14b-47c0-9ede-f2fa4a72ff43)

Just for info, no idea how you could solve your issue, except by hacking these bit of code to see if some hardcoded integer value could be fine - if you need fractional real final size, it's not possible.

**Comment by offset-torque** [comment]
Since this level of fine adjustment is not possible in KOReader **teddympq**, I think you can reach a satisfying solution by changing to a different font. Due to the different font metrics, not all 18 pt fonts look the same. Even at the exact same size, their apparent size can be different due to width and stroke, which sounds like the difference you are looking for.

There are many nice serif/sans serif fonts on the **Google Fonts** website. You can set a small size there and observe the differences. Probably you will find a better fitting one if you are not specifically attached to your current font.

**Comment by jospalau** [comment]
You can create some profiles with different sizes. The profiles let you select sizes by 0.5 steps but you probably can modify the profiles.lua file later to set different values. Then, after applying each profile you can see the font pt size calculated by KOReader as on the previous screenshot. In any case I think the value is Didot points and may differ.

**Comment by teddympq** [comment]
> In the end, it's a size in (integer, no fractional) pixels.

Maybe It would avoid rounding to integers by deleting this specific part?:
`math.ceil`
in https://github.com/koreader/koreader-base/blob/5c7d8bc2b5ca15938880711f75ddc045251e9625/ffi/framebuffer.lua#L412-L424 ?

**Comment by NiLuJe** [comment]
It's an integer because CRe requires integers for its font ramp, not because we thought it's be fun to truncate just for the hell of it ;).

**Comment by teddympq** [comment]
You are right, and I'm sorry if my comments came across as annoying or presumptive, I thought it was rounded to avoid larger steps between fonts size or to avoid other issues regarding antialiasing.
